Sources/AsyncHTTPClient/NoOpLogHandler.swift

```diff
@@ -0,0 +1,40 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2020 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+
+internal struct NoOpLogHandler: LogHandler {
+    func log(level: Logger.Level, message: Logger.Message, metadata: Logger.Metadata?, file: String, function: String, line: UInt) {}
+
+    subscript(metadataKey _: String) -> Logger.Metadata.Value? {
+        get {
+            return nil
+        }
+        set {}
+    }
+
+    var metadata: Logger.Metadata {
+        get {
+            return [:]
```

Comment:
Dictionaries, Arrays etc are _value_ types in Swift, so the programmer doesn't control the references for the backing storage. Because it's a value type however, the programmer of the data structure can be smart (and Array & Dictionary are) and use a single backing reference for the "empty" value.

You may have seen `_emptyArrayStorage` before, that is that already. So under the hood, it does everything you suggest already, without needing to do caching of the empty value ourselves :).

Reply:
Good to know!